v-next/hardhat/src/internal/cli/main.ts

```diff
@@ -265,3 +269,192 @@ function getTaskFromCliArguments(
+    const paramInfo = task.namedParameters.get(paramName);
+
+    if (paramInfo === undefined) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.UNRECOGNIZED_NAMED_PARAM,
+        {
+          parameter: arg,
+        },
+      );
+    }
+
+    usedCliArguments[i] = true;
+
+    if (paramInfo.parameterType === ParameterType.BOOLEAN) {
+      if (
+        usedCliArguments[i + 1] === false &&
+        ["true", "false"].includes(cliArguments[i + 1])
+      ) {
+        // The flag could be follow by the boolean value
+        taskArguments[paramName] = formatParameterValue(
+          cliArguments[i + 1],
+          ParameterType.BOOLEAN,
+          paramName,
+        );
+
+        usedCliArguments[i + 1] = true;
+        continue;
+      }
+
+      // If the flag is not followed by a boolean value, then the flag itself means that the value is true
+      taskArguments[paramName] = true;
+      continue;
+    }
+
+    // The value immediately following a named parameter (if the parameter does not behave as a flag)
+    // is the parameter's value; otherwise, it's an error
+    if (usedCliArguments[i + 1] === false) {
+      i++;
+
+      taskArguments[paramName] = formatParameterValue(
+        cliArguments[i],
+        paramInfo.parameterType,
+        paramName,
+      );
+
+      usedCliArguments[i] = true;
+
+      continue;
+    }
+
+    throw new HardhatError(
+      HardhatError.ERRORS.ARGUMENTS.MISSING_VALUE_FOR_NAMED_PARAMETER,
+      {
+        paramName: arg,
+      },
+    );
+  }
+}
+
+function parsePositionalAndVariadicParameters(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+  taskArguments: Record<string, any>,
+) {
+  let paramI = 0;
+
+  for (let i = 0; i < cliArguments.length; i++) {
+    if (usedCliArguments[i] === true) {
+      continue;
+    }
+
+    if (cliArguments[i] === "--") {
+      // A standalone '--' is ok because it is used to separate CLI tool arguments from script arguments,
+      // ensuring the tool passes subsequent options directly to the script.
+      usedCliArguments[i] = true;
+      continue;
+    }
+
+    const paramInfo = task.positionalParameters[paramI];
+
+    if (paramInfo === undefined) {
+      continue;
+    }
+
+    usedCliArguments[i] = true;
+
+    const formattedValue = formatParameterValue(
+      cliArguments[i],
+      paramInfo.parameterType,
+      paramInfo.name,
+    );
+
+    if (paramInfo.isVariadic === false) {
+      taskArguments[task.positionalParameters[paramI++].name] = formattedValue;
+      continue;
+    }
+
+    // Handle variadic parameters
+    taskArguments[paramInfo.name] = taskArguments[paramInfo.name] ?? [];
+    taskArguments[paramInfo.name].push(formattedValue);
+  }
+}
+
+function kebabToCamelCase(str: string) {
+  return str.replace(/-./g, (match) => match.charAt(1).toUpperCase());
+}
+
+function formatParameterValue(
+  value: string,
+  type: ParameterType,
+  name: string,
+): any {
```

Comment:
Instead of any, we can properly type this. It can be done in a follow-up PR. @schaable already worked on something similar for the task builders.